### PR TITLE
[Torch] [TMTensor] Added mask and is_causal support for torch.aten.scaled_dot_product_attention

### PR DESCRIPTION
**Pull request by rohan-tan-bhowmik**
Enabled mask and is_causal parameters for torch.aten.scaled_dot_product attention + relevant comments + tests.

The tests added highlight the new capabilities introduced in this PR, including:

Attention with F16 mask
Attention with Boolean mask
Causal attention with same Q K V shapes
Causal attention without Q K V shapes

Made sure that one cannot input both mask and is_causal.

### CODE REVIEW

File: projects/pt1/e2e_testing/xfail_sets.py [hunk 1]
```diff
@@ -498,7 +503,12 @@
     "ViewCollapseDynamicWithAtenSizeIntModule_basic",
     "ViewSizeFromOtherTensor_basic",
     "WeightNormInterfaceModule_basic",
+    # REMOVE ON WHEN ENABLE_GQA IS ADDED
```
Comment (raikonenfnu) [inline projects/pt1/e2e_testing/xfail_sets.py:506]: Isn't enable_gqa already added?

Comment (rohan-tan-bhowmik) [reply projects/pt1/e2e_testing/xfail_sets.py:506]: Not on stable (v2.4.0), it works on v2.5.0

Comment (raikonenfnu) [reply projects/pt1/e2e_testing/xfail_sets.py:506]: Can we not just change the test to use `attn_mask=mask` then we don't need to care about GQA for now.

Comment (rohan-tan-bhowmik) [reply projects/pt1/e2e_testing/xfail_sets.py:506]: It's expecting GQA while torch doesn't have it yet, we need to wait for that

File: lib/Dialect/TMTensor/IR/TMTensorOps.cpp [hunk 2]
```diff
@@ -214,14 +256,54 @@ LogicalResult AttentionOp::generateScalarImplementation(OpBuilder &b,
...
+      SmallVector<Value>(weightRank, one),
+      [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+        Value x = b.create<memref::LoadOp>(loc, weight, localIVs);
+        x = b.create<arith::DivFOp>(loc, x, scaleFactor);
+        b.create<memref::StoreOp>(loc, x, weight, localIVs);
+      });
+
+  // Apply mask to weights if mask is given
+  if (mask) {
+    if (maskType.getElementType().isInteger(1)) {
```
Comment (rsuderman) [inline lib/Dialect/TMTensor/IR/TMTensorOps.cpp:280]: Move this conditional inside of the `scf::ParallelOp` builder. It limits the difference to the only region that varies (add vs select).

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: projects/pt1/e2e_testing/xfail_sets.py [hunk 3]
```diff
@@ -2178,7 +2190,14 @@
 
 if torch_version_for_comparison() < version.parse("2.5.0.dev"):
     MAKE_FX_TOSA_PASS_SET = MAKE_FX_TOSA_PASS_SET | {
+        # REMOVE ON WHEN ENABLE_GQA IS ADDED
+        "ScaledDotProductAttentionBoolMaskModule_basic",
```
[the remark this answers is not in the document]
Comment (rohan-tan-bhowmik) [reply projects/pt1/e2e_testing/xfail_sets.py]: Addressed

File: lib/Dialect/TMTensor/IR/TMTensorOps.cpp (hunk 2 above)
Comment (rohan-tan-bhowmik) [reply lib/Dialect/TMTensor/IR/TMTensorOps.cpp]: Makes sense

File: lib/Dialect/TMTensor/IR/TMTensorOps.cpp [hunk 4]
```diff
@@ -180,6 +220,8 @@ LogicalResult AttentionOp::generateScalarImplementation(OpBuilder &b,
 
   Value zeroF = b.create<arith::ConstantOp>(loc, elementType,
                                             b.getFloatAttr(elementType, 0.0));
+  Value negInfF = b.create<arith::ConstantOp>(
+      loc, elementType, b.getFloatAttr(elementType, -INFINITY));
```
[the remark this answers is not in the document]
Comment (rohan-tan-bhowmik) [reply lib/Dialect/TMTensor/IR/TMTensorOps.cpp]: Sounds good

File: lib/Dialect/TMTensor/IR/TMTensorOps.cpp [hunk 5]
```diff
@@ -168,10 +203,15 @@ LogicalResult AttentionOp::generateScalarImplementation(OpBuilder &b,
...
   Value value = getValue();
+
+  auto optionalMask = getAttnMask();
+  Value mask = optionalMask ? *optionalMask : Value();
+
   Value output = getOutput();
   auto queryType = cast<MemRefType>(query.getType());
   auto keyType = cast<MemRefType>(key.getType());
   auto valueType = cast<MemRefType>(value.getType());
+  auto maskType = mask ? cast<MemRefType>(mask.getType()) : MemRefType();
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply lib/Dialect/TMTensor/IR/TMTensorOps.cpp:214]: Then you can use `dyn_cast_or_null`

**Comment by rsuderman** [comment]
You need to still add the passing sdpa ops to the stable hlo tests

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: projects/pt1/e2e_testing/xfail_sets.py [hunk 6]
```diff
@@ -4548,7 +4582,14 @@
     "ScalarConstantTupleModule_basic",
     "ScalarImplicitFloatModule_basic",
     "ScalarImplicitIntModule_basic",
-    "ScaledDotProductAttentionSameModule_basic",
+    # REMOVE WHEN ENABLE_GQA IS ADDED
+    "ScaledDotProductAttentionBoolMaskModule_basic",
+    "ScaledDotProductAttentionDifferentCausalModule_basic",
+    # "ScaledDotProductAttentionDifferentModule_basic", (PASSES)
```
[the remark this answers is not in the document]
Comment (rohan-tan-bhowmik) [reply projects/pt1/e2e_testing/xfail_sets.py]: All removed now

**Review by rsuderman — APPROVED** [review]

**Pull request merged**